http-clients/src/main/java/org/triplea/http/client/web/socket/WebSocketConnection.java

```diff
@@ -95,53 +109,71 @@ void close() {
-    try {
-      client.connect();
-      pingSender.start();
-    } catch (final Exception e) {
-      pingSender.cancel();
-      throw new CouldNotConnect(serverUri);
-    }
+
+    connectAsync()
+        .whenComplete(
+            (connected, throwable) -> {
+              if (connected && throwable == null) {
+                pingSender.start();
+              } else {
+                errorHandler.accept("Failed to connect to: " + serverUri);
+              }
+            });
+  }
+
+  // Suppression: Ignoring future return values ignores any exceptions thrown by the code
+  // that completes the future. Connect blocking does not throw any exceptions, there are
+  // no exceptions to ignore, we can ignore the future returned by 'submit'.
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private CompletableFuture<Boolean> connectAsync() {
+    final CompletableFuture<Boolean> completableFuture = new CompletableFuture<>();
+    // execute the connection attempt
+    Executors.newSingleThreadExecutor()
+        .submit(
+            () -> {
+              boolean connected;
+              try {
+                connected =
+                    client.connectBlocking(DEFAULT_CONNECT_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS);
+              } catch (final InterruptedException ignored) {
+                connected = false;
+              }
+              return completableFuture.complete(connected);
+            });
+    return completableFuture;
   }
 
   /**
-   * Sends a message asynchronously.
+   * Sends a message asynchronously. Messages are queued until a connection has been established.
    *
-   * @throws IllegalStateException If the connection hasn't been opened yet. {@link #connect()}
-   *     needs to be called first.
+   * @throws IllegalStateException If the connection has been closed.
    */
   void sendMessage(final String message) {
     Preconditions.checkState(!closed);
-    if (!client.isOpen()) {
-      try {
-        Awaitility.await()
-            .atMost(connectTimeoutMillis, TimeUnit.MILLISECONDS)
-            .until(client::isOpen);
-      } catch (final ConditionTimeoutException ignored) {
-        throw new CouldNotConnect(serverUri);
-      }
-    }
-    client.send(message);
-  }
-
-  /** Exception indicating connection to server failed. */
-  @VisibleForTesting
-  static final class CouldNotConnect extends RuntimeException {
-    private static final long serialVersionUID = -5403199291005160495L;
 
-    private static final String ERROR_MESSAGE = "Error, could not connect to server at %s";
-
-    CouldNotConnect(final URI uri) {
-      super(String.format(ERROR_MESSAGE, uri));
+    // Synchronized to make sure that the connection does not open right after we check it.
+    // If the connection is not yet open, the synchronized block should guarantee that we add
+    // to the queued message queue before we start flushing it.
+    synchronized (queuedMessages) {
+      if (!client.isOpen()) {
```

Comment:
I wonder if the value of `client.isOpen()` changes before or after `WebSocketClient#onOpen` is run.
Regardless there is a (really) unlikely race-condition here because the isOpen state isn't synchronized as well.

I'd recommend introducing a local field open that tracks this state independent of implementation

Reply:
`isOpen` is set before `onOpen` is called. For reference, the 'isOpen' value comes from `readyState` when `readyState == ReadyState.OPEN`

***WebsocketImpl.java***
```
	private void open( Handshakedata d ) {
		log.trace( "open using draft: {}",draft );
		readyState = ReadyState.OPEN;
		try {
			wsl.onWebsocketOpen( this, d );
		} catch ( RuntimeException e ) {
			wsl.onWebsocketError( this, e );
		}
	}
```

***WebsocketClient.java***
```
	@Override
	public final void onWebsocketOpen( WebSocket conn, Handshakedata handshake ) {
		startConnectionLostTimer();
		onOpen( (ServerHandshake) handshake );
		connectLatch.countDown();
	}
```

> Regardless there is a (really) unlikely race-condition here because the isOpen state isn't synchronized as well.

The purpose of the synchronization blocks is to avoid any race-condition. I'm curious where you are spotting such a situation.

If we get to the point of code where we start flushing messages, then `client.isOpen` will be true, because we have a lock at that point, we should be blocked from getting into this code block and when we do, `client.isOpen` will evaluate to true and we will not add messages to the queue.

I think I'm happy trusting the implementation on this to not change the behavior `isOpen` being toggled before `onOpen` is called. That seems like a drastic change for them, also strikes me a bit as a bit too much untrusting of the library.

Reply:
The race condition:
Given 2 Threads A and B.
A is the UI (or UI triggered) Thread and B the Thread that actually handles the TCP logic.
So Thread B is performing a websocket handshake and executes until the point where the state is set to `ReadyState.OPEN`. Then the scheduler decides to pause Thread B and resumes with Thread A which is about to send/queue a message. Thread A reaches the synchronized block, checks if `client.isOpen()` which returns true at this point and the message is being sent directly (likely without an exception if I'm not missing something). Then Thread A is done and is stopped.
The scheduler then starts executing Thread B again, enters the synchronized block of the `onOpen` method, sends all existing messages and clears the queue afterwards.
Note that this probably sends all messages correctly, however possibly out of order which could in theory cause some weird issues.
That's why I suggested keeping track of the state ourselves, so we can set the state while synchronized to `queuedMessages`.
However I realized that this might not even be necessary.
If we checked `client.isOpen() && queuedMessages.isEmpty()` instead, we would avoid sending messages out of order as long as the `isOpen` state is being set before the `onOpen` method is run.
Note that in this case it is important to use `isEmpty()` over `size() == 0`, [the javadocs state](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ConcurrentLinkedQueue.html#size()):
> Beware that, unlike in most collections, this method is NOT a constant-time operation. Because of the asynchronous nature of these queues, determining the current number of elements requires an O(n) traversal. Additionally, if elements are added or removed during execution of this method, the returned result may be inaccurate. Thus, this method is typically not very useful in concurrent applications.

Whereas no such comment is found on the `isEmpty()` method.
Out of curiosity I checked the implementation and it doesn't depend on size at all, so it's constant time after all.
What you should consider, even though it might not make a difference:
You already synchronize the access to the collection, so you could easily a simple `ArrayDeque` instead of a more complex concurrent implementation to avoid the overhead.